README.md
# NAME

Stenciller::Plugin::ToMojoliciousTest - Create Mojolicious tests from text files parsed with Stenciller

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10+-blue.svg" alt="Requires Perl 5.10+" />
    <a href="https://travis-ci.org/Csson/p5-Stenciller-Plugin-ToMojoliciousTest"><img src="https://api.travis-ci.org/Csson/p5-Stenciller-Plugin-ToMojoliciousTest.svg?branch=master" alt="Travis status" /></a>
    <a href="http://cpants.cpanauthors.org/dist/Stenciller-Plugin-ToMojoliciousTest-0.0103"><img src="https://badgedepot.code301.com/badge/kwalitee/Stenciller-Plugin-ToMojoliciousTest/0.0103" alt="Distribution kwalitee" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Stenciller-Plugin-ToMojoliciousTest%200.0103"><img src="https://badgedepot.code301.com/badge/cpantesters/Stenciller-Plugin-ToMojoliciousTest/0.0103" alt="CPAN Testers result" /></a>
    <img src="https://img.shields.io/badge/coverage-94.0%-yellow.svg" alt="coverage 94.0%" />
    </p>
</div>

# VERSION

Version 0.0103, released 2016-03-24.

# SYNOPSIS

    use Stenciller;

    my $contents = Stenciller->new(filepath => 'path/to/filetoparse.stencil')->transform(
        plugin_name => 'ToMojoliciousTest',
        transform_args => {
            require_in_extra => {
                key => 'is_test',
                value => 1,
                default => 1
            },
        },
        constructor_args => {
            template => 'path/to/template.test',
        },
    );

# DESCRIPTION

Stenciller::Plugin::ToMojoliciousTest is a [Stenciller](https://metacpan.org/pod/Stenciller) plugin that transforms stencils to Mojolicious tests.

If you build your distribution with [Dist::Zilla](https://metacpan.org/pod/Dist::Zilla) the [Dist::Zilla::Plugin::Stenciller::MojoliciousTests](https://metacpan.org/pod/Dist::Zilla::Plugin::Stenciller::MojoliciousTests) plugin is an easy way to use this class.

If the text file looks like this:

    == stencil ==

    --input--
        %= link_to 'Example', 'http://www.example.com/'
        %= form_for '01_test_1'
    --end input--

    --output--
        <a href="http://www.example.com/">Example</a>
        <form action="/01_test_1"></form>
    --end output--

And the template file like this:

    use 5.10.1;
    use strict;
    use warnings;
    use Test::More;
    use Test::Warnings;
    use Test::Mojo::Trim;
    use Mojolicious::Lite;

    use if $ENV{'AUTHOR_TESTING'}, 'Test::Warnings';

    my $test = Test::Mojo::Trim->new;

Then `$contents` will contain this:

        use 5.10.1;
        use strict;
        use warnings;
        use Test::More;
        use Test::Warnings;
        use Test::Mojo::Trim;
        use Mojolicious::Lite;

        use if $ENV{'AUTHOR_TESTING'}, 'Test::Warnings';

        my $test = Test::Mojo::Trim->new;


        # test from line 1 in 01-test.stencil

        my $expected_01_test_1 = qq{    <a href="http://www.example.com/">Example</a>
    <form action="/01_test_1"></form>};

        get '/01_test_1' => '01_test_1';

        $test->get_ok('/01_test_1')->status_is(200)->trimmed_content_is($expected_01_test_1, 'Matched trimmed content in 01-test.stencil, line 1');

        done_testing();

        __DATA__

        @@ 01_test_1.html.ep

            %= link_to 'Example', 'http://www.example.com/'

            %= form_for '01_test_1'

Do note that the generated tests currently are hardcoded to use `trimmed_content_is` from `Test::Mojo::Trim`. This might change in the future.

# SEE ALSO

- [Stenciller](https://metacpan.org/pod/Stenciller)
- [Dist::Zilla::Plugin::Stenciller::ToMojoliciousTests](https://metacpan.org/pod/Dist::Zilla::Plugin::Stenciller::ToMojoliciousTests)

# SOURCE

[https://github.com/Csson/p5-Stenciller-Plugin-ToMojoliciousTest](https://github.com/Csson/p5-Stenciller-Plugin-ToMojoliciousTest)

# HOMEPAGE

[https://metacpan.org/release/Stenciller-Plugin-ToMojoliciousTest](https://metacpan.org/release/Stenciller-Plugin-ToMojoliciousTest)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Stenciller-Plugin-ToMojoliciousTest

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0103 2016-03-24T08:30:56Z UTC
 [Bug Fixes]
 - Now depends on Mojolicious (instead of Mojolicious::Lite), since that's
   where $VERSION is.

 [Stats & Dependency Changes]
 - 2 files changed, 2 insertions(+), 2 deletions(-)
 - (test req) + Mojolicious 6.00
 - (test req) - Mojolicious::Lite

0.0102 2016-03-23T20:57:43Z UTC
 [Requirements]
 - Lower Perl req to 5.10 (and require Stenciller 0.1400, which makes that
   possible).

 [Stats & Dependency Changes]
 - 4 files changed, 19 insertions(+), 16 deletions(-)
 - (run req) ~ Moose 2.00 --> 2.1405
 - (run req) ~ Path::Tiny (any) --> 0.072
 - (run req) ~ Stenciller 0.1300 --> 0.1400
 - (run req) ~ Types::Path::Tiny (any) --> 0.005
 - (run req) ~ perl 5.014000 --> 5.010000
 - (test req) + IO::Handle (any)
 - (test req) + IPC::Open3 (any)
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Cpantesters (any)
 - (dev req) + Badge::Depot::Plugin::Kwalitee (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Badge::Depot::Plugin::Travis (any)
 - (dev req) + Dist::Zilla 5.000
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.007
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0100
 - (dev req) + Dist::Zilla::Plugin::Clean 0.07
 - (dev req) + Dist::Zilla::Plugin::DistIller::MetaGeneratedBy (any)
 - (dev req) + Dist::Zilla::Plugin::Git 2.036
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Contributors 0.017
 - (dev req) + Dist::Zilla::Plugin::InstallRelease 0.008
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped 0.0200
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Dist::Zilla::Plugin::Test::Compile 2.054
 - (dev req) + Dist::Zilla::Plugin::Test::EOF 0.0501
 - (dev req) + Dist::Zilla::Plugin::Test::EOL 0.18
 - (dev req) + Dist::Zilla::Plugin::Test::Kwalitee::Extra v0.2.1
 - (dev req) + Dist::Zilla::Plugin::Test::NoTabs 0.15
 - (dev req) + Dist::Zilla::Plugin::Test::Version 1.05
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::EOL (any)
 - (dev req) + Test::More 0.88
 - (dev req) + Test::NoTabs 1.4
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.151680
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.003001
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.150250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.028
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs (any) --> 0.021
 - (dev req) ~ Dist::Zilla::Plugin::TravisYML (any) --> 1.15
 - (dev req) ~ Pod::Weaver::Section::Badges (any) --> 0.0401
 - (dev req) ~ Test::Warnings (any) --> 0.026
 - (dev req) - Dist::Zilla::Plugin::CheckExtraTests
 - (dev req) - Dist::Zilla::Plugin::ContributorsFromGit
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::Class
 - (dev req) - Dist::Zilla::Plugin::NextRelease
 - (dev req) - Dist::Zilla::Plugin::NextVersion::Semantic
 - (dev req) - Dist::Zilla::Plugin::OurPkgVersion
 - (dev req) - Dist::Zilla::Plugin::PodnameFromClassname
 - (dev req) - Dist::Zilla::Plugin::Prereqs::Plugins
 - (dev req) - Dist::Zilla::Plugin::PreviousVersion::Changelog
 - (dev req) - Dist::Zilla::Plugin::ReversionOnRelease
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) + Dist::Iller 0.1408
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0309

0.0101 2015-11-23T15:34:08Z UTC
 [Documentation]
 - Fixed link to Dist::Zilla::Plugin::Stenciller::MojoliciousTests.

0.0100 2015-11-23T01:32:26Z UTC
 - Initial release


lib/Stenciller/Plugin/ToMojoliciousTest.pm
use 5.10.0;
use strict;
use warnings;

package Stenciller::Plugin::ToMojoliciousTest;

our $VERSION = '0.0104';
# ABSTRACT: Create Mojolicious tests from text files parsed with Stenciller
# AUTHORITY

use Moose;
use Types::Path::Tiny qw/Path/;
with 'Stenciller::Transformer';

has template => (
    is => 'ro',
    isa => Path,
    required => 1,
    coerce => 1,
);

sub transform {
    my $self = shift;
    my $transform_args = shift;

    my @out = $self->init_out($self->stenciller, $transform_args);

    my @stencils = ();

    STENCIL:
    for my $i (0 .. $self->stenciller->max_stencil_index) {

        next STENCIL if $self->should_skip_stencil_by_index($i, $transform_args);

        my $stencil = $self->stenciller->get_stencil($i);
        next STENCIL if $self->should_skip_stencil($stencil, $transform_args);

        push @stencils => $stencil;
    }

    foreach my $stencil (@stencils) {

        my $expected_var = sprintf '$expected_%s', $stencil->stencil_name;

        push @out => sprintf "# test from line %s in %s", $stencil->line_number, $self->stenciller->filepath->basename;

        #warn "-------\n" . $stencil->all_output . "\n-----------<<";
        push @out => sprintf 'my %s = qq{%s};', $expected_var, join ("\n" => $stencil->all_output);

        push @out => sprintf q{get '/%s' => '%s';}, $stencil->stencil_name, $stencil->stencil_name;

        my $test_comment = sprintf q{Matched trimmed content in %s, line %s} => $self->stenciller->filepath->basename, $stencil->line_number;
        push @out => sprintf q{$test->get_ok('/%s')->status_is(200)->trimmed_content_is(%s, '%s');} => $stencil->stencil_name, $expected_var, $test_comment;

    }

    push @out => 'done_testing();';
    push @out => '__DATA__';

    foreach my $stencil (@stencils) {
        push @out => sprintf '@@ %s.html.ep' => $stencil->stencil_name;
        push @out => $stencil->all_input;
    }

    my $content = join "\n\n" => '', @out, '';
    $content =~s{\v{2,}}{\n\n}g;

    my $template = $self->template->slurp_utf8;
    return $template.$content;
}

1;


__END__

=pod

=head1 SYNOPSIS

    use Stenciller;

    my $contents = Stenciller->new(filepath => 'path/to/filetoparse.stencil')->transform(
        plugin_name => 'ToMojoliciousTest',
        transform_args => {
            require_in_extra => {
                key => 'is_test',
                value => 1,
                default => 1
            },
        },
        constructor_args => {
            template => 'path/to/template.test',
        },
    );

=head1 DESCRIPTION

Stenciller::Plugin::ToMojoliciousTest is a L<Stenciller> plugin that transforms stencils to Mojolicious tests.

If you build your distribution with L<Dist::Zilla> the L<Dist::Zilla::Plugin::Stenciller::MojoliciousTests> plugin is an easy way to use this class.

If the text file looks like this:

    == stencil ==

    --input--
        %= link_to 'Example', 'http://www.example.com/'
        %= form_for '01_test_1'
    --end input--

    --output--
        <a href="http://www.example.com/">Example</a>
        <form action="/01_test_1"></form>
    --end output--

And the template file like this:

    use 5.10.1;
    use strict;
    use warnings;
    use Test::More;
    use Test::Warnings;
    use Test::Mojo::Trim;
    use Mojolicious::Lite;

    use if $ENV{'AUTHOR_TESTING'}, 'Test::Warnings';

    my $test = Test::Mojo::Trim->new;


Then C<$contents> will contain this:

        use 5.10.1;
        use strict;
        use warnings;
        use Test::More;
        use Test::Warnings;
        use Test::Mojo::Trim;
        use Mojolicious::Lite;

        use if $ENV{'AUTHOR_TESTING'}, 'Test::Warnings';

        my $test = Test::Mojo::Trim->new;


        # test from line 1 in 01-test.stencil

        my $expected_01_test_1 = qq{    <a href="http://www.example.com/">Example</a>
    <form action="/01_test_1"></form>};

        get '/01_test_1' => '01_test_1';

        $test->get_ok('/01_test_1')->status_is(200)->trimmed_content_is($expected_01_test_1, 'Matched trimmed content in 01-test.stencil, line 1');

        done_testing();

        __DATA__

        @@ 01_test_1.html.ep

            %= link_to 'Example', 'http://www.example.com/'

            %= form_for '01_test_1'

Do note that the generated tests currently are hardcoded to use C<trimmed_content_is> from C<Test::Mojo::Trim>. This might change in the future.

=head1 SEE ALSO

=for :list
* L<Stenciller>
* L<Dist::Zilla::Plugin::Stenciller::ToMojoliciousTests>

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.043.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Create Mojolicious tests from text files parsed with Stenciller",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Stenciller-Plugin-ToMojoliciousTest",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Stenciller::Plugin::ToMojoliciousTest",
  "PREREQ_PM" => {
    "Moose" => "2.1405",
    "Path::Tiny" => "0.072",
    "Stenciller" => "0.1400",
    "Types::Path::Tiny" => "0.005"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "IPC::System::Simple" => 0,
    "Mojolicious" => "6.00",
    "String::Cushion" => 0,
    "Syntax::Feature::Qi" => 0,
    "Test::Differences" => 0,
    "Test::Mojo::Trim" => 0,
    "Test::More" => "0.96",
    "syntax" => 0
  },
  "VERSION" => "0.0104",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "IPC::System::Simple" => 0,
  "Mojolicious" => "6.00",
  "Moose" => "2.1405",
  "Path::Tiny" => "0.072",
  "Stenciller" => "0.1400",
  "String::Cushion" => 0,
  "Syntax::Feature::Qi" => 0,
  "Test::Differences" => 0,
  "Test::Mojo::Trim" => 0,
  "Test::More" => "0.96",
  "Types::Path::Tiny" => "0.005",
  "syntax" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
